File: app/javascript/channels/room_channel.js
/* global $*/
import consumer from "./consumer"

const channel_name = "RoomChannel"

$(document).on("turbolinks:load",function(){
  const data = $("#data").data();
  if(data==null){
    // delete subscriptions
    const subscriptions = consumer.subscriptions.subscriptions;
    if(subscriptions.length > 0){
      subscriptions.map((subscription)=>{subscription.unsubscribe()})
    };
  } else {
    const room_id = data.roomId;
    const user_id = data.userId;
    if(isSubscribed(room_id,user_id)){return};

    // make subscription
    appRoom(room_id, user_id);
  };
});

const appRoom = (room_id, user_id) => consumer.subscriptions.create(
  {channel: channel_name, room_id: room_id, user_id: user_id}, {
  connected() {
  },

  disconnected() {
  },

  received(data) {
    console.log(data);
    if(data['start']){window.location.href = data['start'];};
    console.log(data['master']);
    console.log(data['master']==user_id);
    console.log(data['master']===user_id);
    if(data['master']==user_id){$("#start-button").show()} else {$("#start-button").hide()};
    $("#members").html(data['member']);
    $("#results").html(data['result']);
    $("#player-"+data['player_id']).html(data['player']);
    $("#answer-"+data['player_id']).html(data['answer'])
  }
});

const isSubscribed = (room_id, user_id) => {
  const identifier = `{"channel":"${channel_name}","room_id":${room_id},"user_id":${user_id}}`
  const subscription = consumer.subscriptions.findAll(identifier)
  return !!subscription.length
}

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel

  def subscribed
    @user = User.find(params[:user_id])
    reject if @user.nil?
    @room = Room.find(params[:room_id])
    reject if @room.nil?
    stream_for(@room)
    puts "---------create subscription -----------"
    join_room(params[:room_id], params[:user_id])
  end

  def unsubscribed
    stop_all_streams
    puts "---------delete subscription -----------"
    left_room(params[:room_id], params[:user_id])
  end

private

  def join_room(room_id, user_id)
    room = Room.find(room_id)
    room.user_rooms.create!(user_id: user_id) unless room.user_rooms.exists?(user_id: user_id)
    broadcast_member(room, user_id)
  end

  def left_room(room_id, user_id)
    room = Room.find(room_id)
    connection = room.user_rooms.find_by(user_id: user_id)
    connection.destroy if connection.present?
    broadcast_member(room, user_id)
  end

  def broadcast_member(room, user_id)
    member_template = ApplicationController.renderer.render partial: 'rooms/members', locals: { users: room.users }
    player_template = ApplicationController.renderer.render partial: 'users/online', locals: { player: User.find(user_id), room: room}
    master = room.user_rooms.first
    master_id = master ? master.user_id : 0
    RoomChannel.broadcast_to(room, {member: member_template, master: master_id, player_id: user_id, player: player_template})
  end

end
